tracer/src/Datadog.Trace/Propagators/SpanContextPropagator.cs

```diff
@@ -150,76 +154,101 @@ internal void Inject<TCarrier, TCarrierSetter>(SpanContext context, TCarrier car
         }
 
-        internal SpanContext? Extract<TCarrier, TCarrierGetter>(TCarrier carrier, TCarrierGetter carrierGetter)
+        internal PropagationContext Extract<TCarrier, TCarrierGetter>(TCarrier carrier, TCarrierGetter carrierGetter)
             where TCarrierGetter : struct, ICarrierGetter<TCarrier>
         {
             if (carrier is null) { ThrowHelper.ThrowArgumentNullException(nameof(carrier)); }
 
-            SpanContext? localSpanContext = null;
+            // as we extract values from the carrier using multiple extractors,
+            // we will accumulate them in this context
+            SpanContext? cumulativeSpanContext = null;
+            Baggage? cumulativeBaggage = null;
 
-            for (var i = 0; i < _extractors.Length; i++)
+            foreach (var extractor in _extractors)
             {
-                if (_extractors[i].TryExtract(carrier, carrierGetter, out var spanContext))
+                if (_propagationExtractFirstOnly &&
+                    cumulativeSpanContext is not null &&
+                    extractor.PropagatorType == PropagatorType.TraceContext)
                 {
-                    if (_propagationExtractFirstOnly)
-                    {
-                        return spanContext;
-                    }
+                    // skip this extractor if
+                    // - we are configured to extract only the first trace context (aka SpanContext)
+                    // - we already extracted trace context
+                    // - this extractor is a trace context extractor
+                    continue;
+                }
+
+                if (!extractor.TryExtract(carrier, carrierGetter, out var currentExtractedContext))
+                {
+                    // extractor failed to extract
+                    continue;
+                }
 
-                    if (localSpanContext is not null && spanContext is not null && _extractors[i] is W3CTraceContextPropagator)
+                if (cumulativeBaggage is null)
+                {
+                    cumulativeBaggage = currentExtractedContext.Baggage;
+                }
+                else if (currentExtractedContext.Baggage is { Count: > 0 } extractedBaggage)
+                {
+                    cumulativeBaggage.Merge(extractedBaggage);
+                }
+
+                if (cumulativeSpanContext == null)
+                {
+                    cumulativeSpanContext = currentExtractedContext.SpanContext;
+                }
+                else if (extractor is W3CTraceContextPropagator && currentExtractedContext.SpanContext is { } extractedSpanContext)
+                {
+                    FixUpSpanContext(cumulativeSpanContext, extractedSpanContext);
+                }
+            }
+
+            return new PropagationContext(cumulativeSpanContext, cumulativeBaggage);
+        }
+
+        private static void FixUpSpanContext(SpanContext cumulativeSpanContext, SpanContext extractedSpanContext)
```

Comment:
I have two unique ideas for renaming this for clarity. LMK what you think:
1) This included suggestion, if you don't think it's too verbose
```suggestion
        private static void MergeExtractedW3CSpanContext(SpanContext cumulativeSpanContext, SpanContext extractedSpanContext)
```
2) An extension method so we have a `Merge` operation similar to the Baggage one, so `internal static void MergeW3CSpanContext(this SpanContext spanContext, SpanContext w3cSpanContext);`

Reply:
Thanks for the recommendation. I went with just renaming because it's much easier.